crates/turbo-tasks/src/join_iter_ext.rs

```diff
@@ -137,20 +135,29 @@ where
 
 pub trait TryFlatMapRecursiveJoinIterExt<T, C, F, CI>: Iterator
 where
-    T: Hash + PartialEq + Eq + Clone,
-    C: Fn(T) -> F,
+    C: FnMut(&T) -> Option<F>,
     F: Future<Output = Result<CI>>,
     CI: IntoIterator<Item = T>,
 {
-    /// Applies the `get_children` function on each item in the iterator, and on
-    /// each item that is returned by `get_children`. Collects all items from
-    /// the iterator and all items returns by `get_children` into an index set.
-    /// The order of items is equal to a breadth-first traversal of the tree,
-    /// but `get_children` will execute concurrently. It will handle circular
-    /// references gracefully. Returns a future that resolve to a
-    /// [Result<IndexSet>]. It will resolve to the first error that occur in
-    /// breadth-first order.
-    fn try_flat_map_recursive_join(self, get_children: C) -> TryFlatMapRecursiveJoin<T, C, F, CI>;
+    /// Applies the `filter_flat_map` function on each item in the iterator, and
+    /// on each item that is returned by `filter_flat_map`, recursively.
+    ///
+    /// Collects all items from the iterator and all items returns by
+    /// `filter_flat_map` into a vector.
+    ///
+    /// `filter_flat_map` will execute concurrently
+    ///
+    /// **Beware:**
+    /// * The order of the returned items is undefined.
+    /// * Circular references must be handled within `filter_flat_map`: return
+    ///   `None` to stop the recursion.
```

Comment:
TBH I don't feel good about that.

Can we merge the topo sort into that method to make it deterministic?

Having a non-deterministic helper method feels like a source of mistakes...